### ISSUE TITLE
Documentation on how to vmap on dictionary entries

### ISSUE THREAD

**Issue by lukasbrauncom** [google/jax#3161]
I have a related question to #2367: Is it possible to use `vmap` on an array that is in a dictionary? And if yes, how?

To stick with the example given in #2367, how to `vmap` 'b', which is an array within a dictionary:
```
import jax.numpy as np
from jax import vmap

dictionary = {'a': 5., 'b': np.arange(5)}
c = 1.
d = 2.

def f(dct, c, d):
  return dct['a'] + dct['b'] + c + d

result = vmap(f, magic)(dictionary, c, d)
```

I don't understand how I would have to create the magic axes tuple for this example or if it is even possible? I think it would be great if there would be a related example in the [documentation of vmap](https://jax.readthedocs.io/en/latest/jax.html?highlight=vmap#jax.vmap).

**Comment by mattjj** [comment]
Yes, set `magic = ({'a': None, 'b': 0}, None, None)`, like this:

```python
import jax.numpy as np
from jax import vmap

dictionary = {'a': 5., 'b': np.arange(5)}
c = 1.
d = 2.

def f(dct, c, d):
  return dct['a'] + dct['b'] + c + d

result = vmap(f, in_axes=({'a': None, 'b': 0}, None, None))(dictionary, c, d)
```

This sentence in the vmap docstring is intended to describe the behavior:

> If the positional arguments to fun are container types, the corresponding element of in_axes can itself be a matching container, so that distinct array axes can be mapped for different container elements. in_axes must be a container tree prefix of the positional argument tuple passed to fun.

Here, the argument in question is a dictionary, so we make the corresponding entry of `in_axes` a dictionary too (with the same keys).

**Comment by lukasbrauncom** [comment]
Thank you - that helps! Now where I see it, I have to admit that this is actually quite obvious.

The reason why I went down the rabbit hole was this error message, which I got after I directly past a dictionary instead of a tuple containing the dictionary, which made me believe that dictionaries are not allowed in the `in_axes` argument: `TypeError: vmap arguments in_axes and out_axes must each be an integer, None, or a (nested) tuple of those types, got <class 'dict'> and <class 'int'> respectively.`

And then I tried a lot of different ways to represent the dictionary in a tuple. With your new example it should now be clear even for me ;-). Thanks!
